### PR TITLE
Multi Objective CNN benchmark - [WIP]

### PR DESCRIPTION
**Pull request by ayushi-3536**
- Added benchmark outline for CNN benchmarks from [paper](https://arxiv.org/abs/2105.01015)
- Pull from MO interface by Phillip is pending

### CODE REVIEW

File: hpobench/benchmarks/mo/cnn_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+                   fidelity: Union[CS.Configuration, Dict, None] = None,
+                   rng: Union[int, np.random.RandomState, None] = None):
+        """ Function that returns the model initialized based on the configuration and fidelity
+        """
+        rng = self.rng if rng is None else rng
+
+        if isinstance(config, CS.Configuration):
+            config = config.get_dictionary()
+        if isinstance(fidelity, CS.Configuration):
+            fidelity = config.get_dictionary()
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/cnn_benchmark.py]: Passing the fidelity as argument seems have no effect. (not returned afterwards)

File: hpobench/benchmarks/mo/cnn_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+        return Net(config, (3, 16, 16), 17)
+
+    @AbstractBenchmark.check_parameters
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/cnn_benchmark.py]: Please use 

```suggestion
    @AbstractMultiObjectiveBenchmark.check_parameters
```
here. Also, please for the objective function test.

File: hpobench/benchmarks/mo/cnn_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+                images = images.to(device)
+                labels = labels.to(device)
+
+                outputs = self(images)
+                acc = accuracy(labels, outputs)
+
+        return acc
+
+
+class CNNBenchmark(AbstractBenchmark):
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/cnn_benchmark.py]: After rebasing to the new state, we can use the MO interface

```suggestion
class CNNBenchmark(AbstractMultiObjectiveBenchmark):
```

File: hpobench/benchmarks/mo/cnn_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+            ),
+            CS.CategoricalHyperparameter(
+                'kernel_size', default_value=5, choices=[7, 5, 3]
+            )
+
+        ])
+        return cs
+
+    @staticmethod
+    def get_objectives():
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/cnn_benchmark.py]:
```suggestion
    def get_objective_names(self) -> List[str]:
```

File: hpobench/benchmarks/mo/cnn_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+            tst_acc=test_accuracy,
+            len=np.log10(num_params),
+            train_runtime=training_runtime,
+            eval_valid_runtime=eval_valid_runtime,
+            eval_test_runtime=eval_test_runtime,
+
+        )
+        t.close()
+
+        return {'function_value': {'accuracy': val_accuracy,
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/cnn_benchmark.py]: Can you please make sure that the returned vales are python objects. Non-python objects might cause some problems when using with the container. 

Example: `val_accuracy.item()`

File: hpobench/benchmarks/mo/cnn_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+                         'valid_accuracy': val_accuracy,
+                         'valid_cost': eval_valid_runtime,
+                         'test_accuracy': test_accuracy,
+                         'test_cost': eval_test_runtime,
+                         'model_size': num_params,
+                         'fidelity': fidelity
+                         }
+                }
+
+    @AbstractBenchmark.check_parameters
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/cnn_benchmark.py]:
```suggestion
    @AbstractMultiObjectiveBenchmark.check_parameters
```

File: hpobench/util/data_manager.py [hunk 7]
```diff
@@ -845,6 +844,89 @@ def _load(self) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndar
...
+
+        t = time()
+        self._download()
+        X_trn, y_trn, X_val, y_val, X_tst, y_tst = self._load()
+        self.logger.info(f'CNNDataManager: Data successfully loaded after {time() - t:.2f}')
+
+        return X_trn, y_trn, X_val, y_val, X_tst, y_tst
+
+    @lockutils.synchronized('not_thread_process_safe', external=True,
+                            lock_path=f'{hpobench.config_file.cache_dir}/lock_protein_structure_data', delay=0.5)
```
Comment (PhMueller) [inline hpobench/util/data_manager.py]: can you give the lock please another name?

File: hpobench/benchmarks/mo/cnn_benchmark.py [hunk 8]
```diff
@@ -0,0 +1,520 @@
...
+
+            layers.append(nn.Linear(inp_n, out_n))
+            layers.append(nn.ReLU())
+
+            inp_n = out_n
+
+        layers.append(nn.Linear(inp_n, num_classes))
+        self.fc_layers = nn.Sequential(*layers)
+
+        self.time_train = 0
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/cnn_benchmark.py]: This attribute seems to be unused

File: hpobench/benchmarks/mo/cnn_benchmark.py [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+            ),
+            CS.CategoricalHyperparameter(
+                'kernel_size', default_value=5, choices=[7, 5, 3]
+            )
+
+        ])
+        return cs
+
+    @staticmethod
+    def get_objective_names(self) -> List[str]:
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/cnn_benchmark.py]:
```suggestion
    def get_objective_names() -> List[str]:
```

File: hpobench/benchmarks/mo/cnn_benchmark.py [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+            loss = criterion(logits, labels.argmax(-1))
+            loss.backward()
+            optimizer.step()
+
+            acc = accuracy(labels, logits)
+
+        return acc
+
+    def eval_fn(self, loader, device):
+        """
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/cnn_benchmark.py:141]: Could you please use the google style for docstrings.

File: hpobench/benchmarks/mo/cnn_benchmark.py [hunk 11]
```diff
@@ -0,0 +1,520 @@
...
+        fidelity1 = dict(
+            fixed=CS.Constant('budget', value=25),
+            variable=CS.UniformIntegerHyperparameter(
+                'budget', lower=1, upper=25, default_value=25, log=False
+            )
+        )
+
+        budget = fidelity1[iter_choice]
+        return budget
+
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/cnn_benchmark.py]: We could merge this function with the `get_fidelity_space` function to increase readability.

File: hpobench/benchmarks/mo/cnn_benchmark.py [hunk 12]
```diff
@@ -0,0 +1,520 @@
...
+    def get_meta_information() -> Dict:
+        """ Returns the meta information for the benchmark """
+        return {
+            'name': 'Bag of baselines for multi-objective joint neural architecture search and hyperparameter optimization',
+            'references': ['@article{guerrero2021bag,'
+                           'title   = {Bag of baselines for multi - objective joint neural architecture search and hyperparameter optimization},'
+                           'author  = {Guerrero-Viu, Julia and Hauns, Sven and Izquierdo, Sergio and Miotto, Guilherme and Schrodi, Simon and Biedenkapp, Andre and Elsken, Thomas and Deng, Difan and Lindauer, Marius and Hutter, Frank},},'
+                           'journal = {arXiv preprint arXiv:2105.01015},'
+                           'year    = {2021}}',
+                           ],
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/cnn_benchmark.py]: Could you please add line breaks, so that the 120 chars per line limit is fulfilled.

Comment (PhMueller) [reply hpobench/benchmarks/mo/cnn_benchmark.py]: in general, also please check the code with;

``` bash
flake8 --max-line-length=120 ./hpobench
pycodestyle --max-line-length=120 ./hpobench
pylint --disable=all --enable=W0221 ./hpobench
```

File: hpobench/benchmarks/mo/cnn_benchmark.py [hunk 13]
```diff
@@ -0,0 +1,520 @@
...
+            'code': 'https://github.com/automl/multi-obj-baselines',
+        }
+
+    def init_model(self, config: Union[CS.Configuration, Dict],
+                   rng: Union[int, np.random.RandomState, None] = None):
+        """ Function that returns the model initialized based on the configuration and fidelity
+        """
+        rng = self.rng if rng is None else rng
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/cnn_benchmark.py]: `rng` is not used

File: hpobench/util/data_manager.py [hunk 14]
```diff
@@ -845,6 +844,89 @@ def _load(self) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndar
...
+
+        t = time()
+        self._download()
+        X_trn, y_trn, X_val, y_val, X_tst, y_tst = self._load()
+        self.logger.info(f'CNNDataManager: Data successfully loaded after {time() - t:.2f}')
+
+        return X_trn, y_trn, X_val, y_val, X_tst, y_tst
+
+    @lockutils.synchronized('not_thread_process_safe', external=True,
+                            lock_path=f'{hpobench.config_file.cache_dir}/lock_mo_cnn_data', delay=0.5)
```
Comment (PhMueller) [inline hpobench/util/data_manager.py]: Can you please remove this for now. 

We have already some generic functions for the dataloading. But since they create their own lock, we dont need that here. 

(I'll make a rework of the data manager soon. )

Comment (PhMueller) [reply hpobench/util/data_manager.py]:
```suggestion
```

File: hpobench/util/data_manager.py [hunk 15]
```diff
@@ -845,6 +844,89 @@ def _load(self) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndar
...
+        if (self.compressed_data).exists():
+            self.logger.debug('CNNDataManager: Data already downloaded')
+        else:
+
+            self.logger.info(f'CNNDataManager: Start downloading data from {self.url_source} '
+                             f'to {self.save_dir}')
+            self.compressed_data.parent.mkdir(parents=True, exist_ok=True)
+            urlretrieve(self.url_source, self.compressed_data)
+            tar = tarfile.open(self.compressed_data)
+            tar.extractall(self.save_dir)
```
Comment (PhMueller) [inline hpobench/util/data_manager.py]: See comment above. 
Also just as a tip, it is always good to use `file.open(..)` inside a context manager (`with tarfile.open(..)`) otherwise you have to close it afterwards manually (`tarfile.close()`)

```suggestion
            self._download_file_with_progressbar(data_url=self.url_source, data_file=self.compressed_data)
            self._untar_data(compressed_file=self.compressed_data, save_dir=self.save_dir)
```

File: hpobench/benchmarks/mo/cnn_benchmark.py [hunk 16]
```diff
@@ -0,0 +1,520 @@
...
+
+        start = time.time()
+        t = tqdm.tqdm(total=epochs)
+        for epoch in range(epochs):
+            train_accuracy = model.train_fn(optimizer, criterion, ds_train, device).item()
+            t.set_postfix(train_accuracy=train_accuracy)
+            t.update()
+        training_runtime = time.time() - start
+
+        num_params = np.sum(p.numel() for p in model.parameters())
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/cnn_benchmark.py]:
```suggestion
        num_params = np.sum([p.numel() for p in model.parameters()])
```

File: hpobench/benchmarks/mo/cnn_benchmark.py [hunk 17]
```diff
@@ -0,0 +1,520 @@
...
+                Note: the number of epoch is 1 indexed! (Results after the first epoch: epoch = 1)
+
+            Fidelity parameters, check get_fidelity_space(). Uses default (max) value if None.
+        rng : np.random.RandomState, int, None
+            Random seed to use in the benchmark.
+
+            To prevent overfitting on a single seed, it is possible to pass a
+            parameter ``rng`` as 'int' or 'np.random.RandomState' to this function.
+            If this parameter is not given, the default random state is used.
+
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/cnn_benchmark.py]: Add docstring for `shuffle` please

File: hpobench/benchmarks/mo/cnn_benchmark.py [hunk 18]
```diff
@@ -0,0 +1,520 @@
...
+        configuration
+        fidelity: Dict, None
+            epoch: int - Values: [1, 50]
+                Number of epochs an architecture was trained.
+                Note: the number of epoch is 1 indexed. (Results after the first epoch: epoch = 1)
+
+            Fidelity parameters, check get_fidelity_space(). Uses default (max) value if None.
+        rng : np.random.RandomState, int, None
+            Random seed to use in the benchmark.
+
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/cnn_benchmark.py]: see above

File: hpobench/benchmarks/mo/cnn_benchmark.py [hunk 19]
```diff
@@ -0,0 +1,520 @@
...
+        start = time.time()
+        t = tqdm.tqdm(total=epochs)
+        for epoch in range(epochs):
+            train_accuracy = model.train_fn(optimizer, criterion, ds_train, device).item()
+            t.set_postfix(train_accuracy=train_accuracy)
+            t.update()
+        training_runtime = time.time() - start
+
+        num_params = np.sum(p.numel() for p in model.parameters())
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/cnn_benchmark.py]:
```suggestion
        num_params = np.sum([p.numel() for p in model.parameters()])
```

File: hpobench/container/benchmarks/mo/cnn_benchmark.py [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+"""
+
+from hpobench.container.client_abstract_benchmark import AbstractBenchmarkClient
+
+
+class FlowerCNNBenchmark(AbstractBenchmarkClient):
+    def __init__(self, **kwargs):
+        kwargs['benchmark_name'] = kwargs.get('benchmark_name', 'FlowerCNNBenchmark')
+        kwargs['container_name'] = kwargs.get('container_name', 'mo_cnn')
+        kwargs['latest'] = kwargs.get('container_tag', '0.0.4')
```
Comment (PhMueller) [inline hpobench/container/benchmarks/mo/cnn_benchmark.py]: I think it should be 0.0.1, right?

File: hpobench/container/benchmarks/mo/cnn_benchmark.py [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+        super(FlowerCNNBenchmark, self).__init__(**kwargs)
+
+
+class FashioCNNBenchmark(AbstractBenchmarkClient):
+    def __init__(self, **kwargs):
+        kwargs['benchmark_name'] = kwargs.get('benchmark_name', 'FashionCNNBenchmark')
+        kwargs['container_name'] = kwargs.get('container_name', 'mo_cnn')
+        kwargs['latest'] = kwargs.get('container_tag', '0.0.4')
```
Comment (PhMueller) [inline hpobench/container/benchmarks/mo/cnn_benchmark.py]: same here

**Comment by PhMueller** [comment]
@KEggensperger, could you please have a look at it?

File: extra_requirements/mo_cnn.json [hunk 22]
```diff
@@ -0,0 +1,7 @@
+{
+  "mo_cnn": [
+    "tqdm",
```
Comment (KEggensperger) [inline extra_requirements/mo_cnn.json]: version?

File: hpobench/benchmarks/mo/cnn_benchmark.py [hunk 23]
```diff
@@ -0,0 +1,591 @@
...
+
+            loss = criterion(logits, labels.argmax(-1))
+            loss.backward()
+            optimizer.step()
+
+            acc = accuracy(labels, logits)
+
+        return acc
+
+    def eval_fn(self, loader, device):
```
Comment (KEggensperger) [inline hpobench/benchmarks/mo/cnn_benchmark.py]: could you please add types for all signatures?

File: hpobench/benchmarks/mo/cnn_benchmark.py [hunk 24]
```diff
@@ -0,0 +1,591 @@
...
+
+    def init_model(self, config: Union[CS.Configuration, Dict],
+                   rng: Union[int, np.random.RandomState, None] = None):
+        """
+        Function that returns the model initialized based on the configuration and fidelity
+        """
+        rng = self.rng if rng is None else rng
+        if isinstance(config, CS.Configuration):
+            config = config.get_dictionary()
+        return Net(config, self.input_shape, self.output_classes)
```
Comment (KEggensperger) [inline hpobench/benchmarks/mo/cnn_benchmark.py:295]: Shouldn't this take the rng?

File: hpobench/benchmarks/mo/cnn_benchmark.py [hunk 25]
```diff
@@ -0,0 +1,591 @@
...
+            t.set_postfix(train_accuracy=train_accuracy)
+            t.update()
+        training_runtime = time.time() - start
+
+        num_params = np.sum([p.numel() for p in model.parameters()]).item()
+        start = time.time()
+        val_accuracy = model.eval_fn(ds_val, device).item()
+        eval_valid_runtime = time.time() - start
+        start = time.time()
+        test_accuracy = model.eval_fn(ds_test, device).item()
```
Comment (KEggensperger) [inline hpobench/benchmarks/mo/cnn_benchmark.py:422]: Same questions as for the other benchmark: Why spending time on computing test metrics?

**Review by KEggensperger — APPROVED** [review]

File: hpobench/benchmarks/mo/cnn_benchmark.py (hunk 25 above)
Comment (PhMueller) [reply hpobench/benchmarks/mo/cnn_benchmark.py:422]: Good question. Changed it to "training time". 

The eval time should be almost equal for every run, so i think it is more important to report the "training time" instead of the "total time per configuration". 

Thanks for the feedback!
